src/OracleFactory.sol

```diff
@@ -1,83 +1,28 @@
-    /// @notice List of running oracles
-    mapping(address => bool) public oracles;
+contract OracleFactory is ERC1967Factory, Ownable {
+    event NewOracleDeployed(address oracle);
 
-    /// @notice Emitted when a new oracle is deployed
-    event NewOracleDeployed(address oracle, Suite suite);
-
-    /// @notice Deploys a new oracle with the specified distributed key and encryption suite
-    /// @dev Only the Factory owner can deploy a new oracle
-    /// @param _distKey The distributed key previously created by a DKG process
-    /// @return The id and address of the new oracle
-    function deployReencryption_BN254_G1_HGAMAL(
-        G1Point calldata _distKey,
-        address _relayer,
-        uint96 _submissionFee,
-        uint96 _reencryptionFee
-    ) external onlyOwner returns (address) {
-        BN254EncryptionOracle oracle = new BN254EncryptionOracle(
-            _distKey,
-            _relayer,
-            _submissionFee,
-            _reencryptionFee
-        );
-
-        oracles[address(oracle)] = true;
-
-        emit NewOracleDeployed(address(oracle), Suite.BN254_KEYG1_HGAMAL);
-        return address(oracle);
-    }
-
-    function pauseOracle(address _oracle) public onlyOwner {
-        require(oracles[_oracle], "no oracle at this address registered");
-        EncryptionOracle oracle = EncryptionOracle(_oracle);
-        oracle.pause();
-    }
-
-    function unpauseOracle(address _oracle) public onlyOwner {
-        require(oracles[_oracle], "no oracle at this address registered");
-        EncryptionOracle oracle = EncryptionOracle(_oracle);
-        oracle.unpause();
-    }
-
-    function updateRelayer(address _oracle, address _newRelayer)
-        public
-        onlyOwner
-    {
-        require(oracles[_oracle], "no oracle at this address registered");
-        EncryptionOracle oracle = EncryptionOracle(_oracle);
-        oracle.updateRelayer(_newRelayer);
-    }
-
-    function updateSubmissionFee(address _oracle, uint96 _submissionFee)
-        public
-        onlyOwner
-    {
-        require(oracles[_oracle], "no oracle at this address registered");
-        EncryptionOracle oracle = EncryptionOracle(_oracle);
-        oracle.updateSubmissionFee(_submissionFee);
+    constructor(address owner) {
+        _initializeOwner(owner);
     }
 
-    function updateReencryptionFee(address _oracle, uint96 _reencryptionFee)
-        public
-        onlyOwner
-    {
-        require(oracles[_oracle], "no oracle at this address registered");
-        EncryptionOracle oracle = EncryptionOracle(_oracle);
-        oracle.updateReencryptionFee(_reencryptionFee);
+    function _deploy(
+        address implementation,
+        address admin,
+        bytes32 salt,
+        bool useSalt,
+        bytes calldata data
+    ) internal override onlyOwner returns (address) {
+        address oracle =
+            ERC1967Factory._deploy(implementation, admin, salt, useSalt, data);
+
+        emit NewOracleDeployed(oracle);
```

Comment:
Same here - should we think about the longterm scope API  by giving a generic address and a type or do you think it's out of scope and easily fixable later by upgrading ?

Reply:
OracleFactory is not upgradeable. This event does just take an address. BTW, this event used to emit (address, Suite), but I changed this so now the nodes need to make an rpc call to `EncryptionOracle.suite()`.

Actually, I can change this so the _deploy() function calls `EncryptionOracle(oracle).suite()` and emits the value in the event. This would add an assurance that the implementation contract (which is passed in) responds to `suite()` and if it doesn't the deployment tx would fail. This is a bit safer